_posts/2019-05-28-improved-hive-bucketing.md

```diff
@@ -0,0 +1,135 @@
+author: David Phillips
+---
+
+The next release of Presto (version 312) adds support for the more
+flexible bucketing introduced in recent versions of Hive. Specifically,
+it allows any number of files per bucket, including zero. This allows
+inserting data into an existing partition without having to rewrite the
+entire partition, and improves the performance of writes by not requiring
+the creation of files for empty buckets.
+
+# Hive bucketing overview
+
+Hive bucketing is a simple form of hash partitioning. A table is bucketed
+on one or more columns with a fixed number of hash buckets:
+
+```sql
+CREATE TABLE page_views (
+  user_id bigint,
+  page_url varchar,
+  dt date
+)
+WITH (
+  partitioned_by = ARRAY['dt'],
+  bucketed_by = ARRAY['user_id'],
+  bucket_count = 50
+)
+```
+
+The bucketing happens within each partition of the table. In the above
+example, the table is partitioned by date and is declared to have `50` buckets
+using the user ID column. This means that the table will have `50` buckets
+for each date. The assigned bucket for each row is determined by hashing
+the user ID value.
+
+# Original Hive bucketing
+
+Originally, Hive required exactly one file per bucket. The files were named 
+such that the bucket number was implicit based on the file's position within
+the lexicographic ordering of the file names. For example, the following list
+of files represent buckets `0` to `2`, respectively:
+
+```
+00000_0
+00001_0
+00002_0
+```
+
+```
+file0
+file3
+file5
+```
+
+```
+bucketA
+bucketB
+bucketD
+```
+
+The file names are meaningless aside from their ordering with respect to the
+other file names.
+
+# What's the problem?
+
+The original Hive bucketing scheme has a couple of problems:
+
+- Inserting data into the table by adding additional files is not possible.
+  Instead, an insert operation requires rewriting all of the existing files,
+  which can be quite expensive.
+  
+- If the data is sparse, some of the buckets might be empty, but because there
+  must be a file for every bucket, the writer must create an empty file for
+  each bucket. Some file formats, such as ORC, support zero-byte files as empty
+  files. Other formats require writing a file with a valid header and footer.
+  Creating these files adds latency to the write operation, and storing these
+  tiny files is inefficient for file systems like HDFS which are designed for
+  large files.
+
+# Improved Hive bucketing
+
+Newer versions of Hive support a bucketing scheme where the bucket number is
+included in the file name. This is the same naming scheme that Hive has always
+used, thus it is backwards compatible with existing data. The naming convention
+has the bucket number as the start of the file name, and requires that the
+number starts with a `0`.
+
+The following list of files shows what data written by Hive might look like for
+a table with a bucket count of `4`:
+
+```
+000000_0            # bucket 0
+000000_0_copy_1     # bucket 0
```

Comment:
You might want to use a different suffix instead of "copy_x", which seems to convey something about the data being another copy of the base file.

Reply:
I agree it looks strange, but this is what Hive does:
```
hive> create table test (x bigint) clustered by (x) into 5 buckets;
OK

hive> insert into test values (123);
Table default.test stats: [numFiles=1, numRows=1, totalSize=4, rawDataSize=3]
OK

hive> insert into test values (123);
Table default.test stats: [numFiles=2, numRows=2, totalSize=8, rawDataSize=6]
OK

hive> insert into test values (123);
Table default.test stats: [numFiles=3, numRows=3, totalSize=12, rawDataSize=9]
OK

hive> insert into test values (456);
Table default.test stats: [numFiles=4, numRows=4, totalSize=16, rawDataSize=12]
OK

hive> dfs -ls hdfs://hadoop-master:9000/user/hive/warehouse/test;
Found 4 items
-rwxrwxrwt   3 root supergroup          4 2019-05-29 12:53 hdfs://hadoop-master:9000/user/hive/warehouse/test/000000_0
-rwxrwxrwt   3 root supergroup          4 2019-05-29 22:56 hdfs://hadoop-master:9000/user/hive/warehouse/test/000000_0_copy_1
-rwxrwxrwt   3 root supergroup          4 2019-05-29 22:59 hdfs://hadoop-master:9000/user/hive/warehouse/test/000000_0_copy_2
-rwxrwxrwt   3 root supergroup          4 2019-05-29 23:00 hdfs://hadoop-master:9000/user/hive/warehouse/test/000000_0_copy_3

Reply:
Hmm, this is wrong, all the files are `000000`, but they shouldn't be

Reply:
This fixes it:
```
hive> SET hive.enforce.bucketing = true;
```
Although the version of Hive we use for testing creates empty files, so it's not exactly the same. I'm surprised that the bucketing doesn't work by default. I guess that fixed that in a later version.